### PR TITLE
[FlashInfer] Avoid D2H sync for sliding-window lengths

### PR DESCRIPTION
**Pull request by hn1209**
## Motivation

When serving Qwen3.5-122B-A10B with DFlash and the FlashInfer draft
attention backend, profiling showed a long `aten::item` on the target
verification path. The call to `paged_kernel_lens.sum().item()` reads a
CUDA scalar back to the host and introduces a stream synchronization in
this latency-sensitive path.

## Modifications

In the sliding-window paged path, `paged_kernel_lens` is computed as:

```python
paged_kernel_lens = torch.minimum(
    seq_lens,
    sliding_window_size + seq_lens - prefix_lens,
)
```

When a caller leaves `prefix_lens` unset and the spec input carries no
`num_accept_tokens`, `update_sliding_window` fills `prefix_lens` with `seq_lens`
itself, so the expression reduces to `min(seq_lens, sliding_window_size)` --
computable from the existing CPU-side sequence-length mirror without reading
the CUDA reduction result back to the host.

Whether that fill happened is decided by control flow, not by tensor values, so
it is recorded at the fill site and consumed directly:

```python
if prefix_lens is None:
    num_accept_tokens = getattr(spec_info, "num_accept_tokens", None)
    prefix_is_full_seq = num_accept_tokens is None
    prefix_lens = seq_lens if prefix_is_full_seq else seq_lens - num_accept_tokens[...]
else:
    prefix_is_full_seq = False

...

if prefix_is_full_seq and seq_lens_cpu is not None:
    paged_kernel_lens_sum = int(
        torch.clamp(seq_lens_cpu, max=sliding_window_size).sum()
    )
else:
    paged_kernel_lens_sum = paged_kernel_lens.sum().item()
```

Recovering the same fact by comparing the two tensors would introduce a D2H
sync of its own on every call where they do not alias, which is exactly what
this path is trying to remove.

Scope: callers that pass explicit `extend_prefix_lens` (regular extend) keep
the original CUDA reduction. The host path applies to the `prefix_lens=None`
entries -- both `is_target_verify()` call sites and `is_draft_extend_v2()`. Of
those, only draft-extend supplies `num_accept_tokens`, so DFlash and EAGLE
target-verify take the host path.

## Accuracy Tests

This change does not modify model computation or attention indices. It
only changes how the host-side total buffer length is calculated for the
equivalent DFlash sliding-window case.

The change was validated with Qwen3.5-122B-A10B and its DFlash draft
model using TP8 and EP8.

Regression coverage: `runner_dflash_verify_swa_window_edges` in
`test/registered/attention/unittests/swa/test_flashinfer.py` exercises the SWA
target-verify path with `prefix_lens=(1, 4, 9)` against `sliding_window_size=4`,
so requests below, at, and above the window each cross-check the host-computed
sum against the reference implementation.

## Speed Tests and Profiling

Test configuration:

- Hardware: 8x NVIDIA A800
- Model: Qwen3.5-122B-A10B
- Speculative algorithm: DFLASH
- Draft attention backend: FlashInfer
- Tensor parallelism: 8
- Expert parallelism: 8
- Speculative draft tokens: 4
- DFlash block size: 4
- Concurrency: 1
- `SGLANG_ENABLE_OVERLAP_PLAN_STREAM=1`

Results:

- TPOT decreased from 5.4 ms to 5.1 ms.
- Before the change, the profiler showed a long `aten::item` /
  `aten::_local_scalar_dense` / `cudaStreamSynchronize` section in target
  verification.
- After the change, the `aten::item` section was no longer present in the
  profile.

**Before**
image

**After**
image

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #33545191233](https://github.com/sgl-project/sglang/actions/runs/33545191233)
Latest PR Test (Extra): :x: [Run #33545188998](https://github.com/sgl-project/sglang/actions/runs/33545188998)
Latest PR Test (AMD ROCm 7.2): :hourglass_flowing_sand: [Run #33545190416](https://github.com/sgl-project/sglang/actions/runs/33545190416)

## Acknowledgements
This work was completed during my employment at Spirit AI as part of an effort to optimize DFlash speculative decoding performance in SGLang.

### CODE REVIEW

**Comment by hn1209** [comment]
Hi maintainers, this is a small change (1 file, +9/-1) in the FlashInfer
backend that removes a device-to-host sync on the sliding-window paged
path. When `prefix_lens == seq_lens`, `paged_kernel_lens` reduces to
`min(seq_lens, sliding_window_size)`, so the sum can be computed from the
existing CPU-side `seq_lens_cpu` mirror instead of reading the CUDA
reduction back to the host. This applies to any case hitting that path
(the DFLASH draft case is one common example); all other cases keep the
original CUDA computation. Could someone add the `run-ci` label to
trigger CI and help kick off the review? Happy to address any feedback.
Thanks!

**Comment by hn1209** [comment]
Hi @merrymercy @Fridge003 @ispobock, this is a small change (1 file, +9/-1) 
in the FlashInfer backend that removes a device-to-host sync on the 
sliding-window paged path. When `prefix_lens == seq_lens`, `paged_kernel_lens` 
reduces to `min(seq_lens, sliding_window_size)`, so the sum can be computed 
from the existing CPU-side `seq_lens_cpu` mirror instead of reading the CUDA 
reduction back to the host. All other cases keep the original computation.

Could someone help add the `run-ci` label to trigger CI and take a look when 
you have time? Happy to address any feedback. Thanks!

File: python/sglang/srt/layers/attention/flashinfer_backend.py [hunk 1]
```diff
@@ -1900,7 +1900,15 @@ def update_sliding_window(
                         seq_lens,
                         sliding_window_size + seq_lens - prefix_lens,
                     )
-                    paged_kernel_lens_sum = paged_kernel_lens.sum().item()
+                    # Avoid D2H sync when prefix_lens defaults to seq_lens (DFLASH draft case)
+                    # In that case: min(seq_lens, window+seq_lens-seq_lens) = min(seq_lens, window)
+                    # which can be computed on CPU using seq_lens_cpu
+                    if seq_lens_cpu is not None and torch.equal(prefix_lens, seq_lens):
```
Comment (hnyls2002) [inline python/sglang/srt/layers/attention/flashinfer_backend.py]: This `torch.equal` still introduces the sync here. The reason you got an improvement was that these two tensors are actually the same object here, which would go through Torch's fast path.

The real fix should utilize the information about whether the prefix lens is None or not. This information is naturally present on the host.

**Comment by hnyls2002** [comment]
/rerun-test test/registered/attention/unittests/swa/test_flashinfer.py

**Comment by hnyls2002** [comment]
/rerun-test test/registered/spec/test_gemma4_dflash_31b_extra.py test/registered/spec/test_gemma4_mtp_31b_extra.py test/registered/spec/dflash/test_dflash.py test/registered/core/test_basic_sanity_dflash.py test/registered/unit/spec/test_dflash_overlap_hostsync.py
